server/src/instant/db/permissioned_transaction_new.clj

```diff
@@ -0,0 +1,410 @@
+    (for [{:keys [op eid etype value rev-etype]} tx-steps
+          :let [key         {:eid eid :etype etype}
+                entity      (get entities-map key)
+                rule-params (get rule-params-map key)]
+          check (cond
+                  (= :add-attr op)
+                  [{:scope    :attr
+                    :action   :create
+                    :etype    "attrs"
+                    :program  (or (rule-model/get-program! rules "attrs" "create")
+                                  {:result true})
+                    :bindings {:data value}}]
+
+                  (and (#{:add-triple :deep-merge-triple :retract-triple} op)
+                       (not entity)) ;; create
+                  (concat
+                   [{:scope    :object
+                     :action   :create
+                     :etype    etype
+                     :eid      (get create-lookups-map eid eid)
+                     :program  (or (rule-model/get-program! rules etype "create")
+                                   {:result true})
+                     :bindings (let [updated-entity (-> (get updated-entities-map key)
+                                                        (update "id" #(get create-lookups-map % %)))]
+                                 {:data        updated-entity
+                                  :new-data    updated-entity
+                                  :rule-params rule-params})}]
+                   ;; updating a ref adds implicit "view" check in reverse direction
+                   ;; with rule-params from forward direction
+                   (when rev-etype
+                     (when-some [rev-entity (get entities-map {:eid value :etype rev-etype})]
+                       [{:scope    :object
+                         :action   :view
+                         :etype    rev-etype
+                         :eid      value
+                         :program  (or (rule-model/get-program! rules rev-etype "view")
+                                       {:result true})
+                         :bindings {:data        rev-entity
+                                    :rule-params rule-params}}])))
+
+                  :else
+                  [])]
+      check)))
+
+(defn run-checks!
+  "Runs checks, returning results (admin-check?) or throwing"
+  [ctx checks]
+  (let [checks (distinct checks)
+        ctx    (assoc ctx :preloaded-refs (cel/create-preloaded-refs-cache))
+        _      (doseq [check checks
+                       :let [{:keys [eid]} check]
+                       :when (lookup-ref? eid)]
+                 (ex/throw-validation-err! :lookup eid [{:message "Could not find the entity for this lookup"}]))
+        results (cel/eval-programs! ctx checks)]
+    (doall
+     (for [{:keys [scope etype result] :as check} results]
+       (do
+         (when-not (:admin-check? ctx)
+           (ex/assert-permitted! :perms-pass? [etype scope] result))
+         (-> check
```

Comment:
Before we had a `data` key with `original` and `updated` in the result. Now we return `:bindings` with `:data` and `:new-data`. The sandbox on the dashboard won't display the results properly if it doesn't get the data in that shape.

Reply:
Fixed the sandbox. New format uses the same names that actual CEL programs use (does not invent new concepts), so I’d like to keep it